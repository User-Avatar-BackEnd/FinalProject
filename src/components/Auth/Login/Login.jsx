import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { Formik, Form } from 'formik';
import * as Yup from 'yup';
import AuthInput from '../AuthInput/AuthInput';
import ErrorMessage from '../../ErrorMessage/ErrorMessage';
import API from '../../../config/API';
import errorsDescription from '../../../config/APIErrorsDescription';

import styles from './Login.module.scss';

const LoginSchema = Yup.object().shape({
  email: Yup.string().email('Invalid email').required('Required'),
  password: Yup.string()
    .strict(false)
    .trim()
    .min(5, 'Must be 5 characters or more')
    .max(64, 'Must be 64 characters or less')
    .matches(/^[a-zA-Z0-9_.-]+$/, 'Can only contain letters, numbers and underscores')
    .required('Required'),
});

const Login = ({ onLogin }) => {
  const [serverError, setServerError] = useState('')

  const clearError = () => {
    setServerError('')
  }

  return (
    <div className={styles.Login}>
      <h2>Sign in</h2>
      <Formik
        initialValues={{ email: '', password: '' }}
        validationSchema={LoginSchema}
        onSubmit={(values) => {
          const data = {
            email: values.email,
            password: values.password
          }

          API.post('/auth/login', data)
            .then(response => {
              onLogin(response.data.access_token)
            })
            .catch(error => {
              setServerError(
                errorsDescription[error.response.data] ?? errorsDescription.default
              )
            })
        }}
      >
        {({ isSubmitting }) => (
          <Form onChange={clearError}>
            <AuthInput type='email' id='email' name='email' label='Email' placeholder='Email' required={true} autoComplete='off' />
            <AuthInput type='password' id='password' name='password' label='Password' placeholder='Password' required={true} />
            <button type='submit' disabled={isSubmitting}>
              Submit
            </button>
          </Form>
        )}
      </Formik>
      <div className={styles.registerButton}>
        <Link to={'/registration'} className={styles.registerLink}>Don't have an account yet? Register</Link>
      </div>
      <ErrorMessage text={serverError} type={'form'} />
    </div>
  );

}

export default Login;