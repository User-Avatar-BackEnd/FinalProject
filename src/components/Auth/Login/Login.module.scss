.Login {
  width: 70%;
  margin: 0 auto;

  & form {
    display: flex;
    flex-direction: column;

    & input {

    }

    & button[type=submit] {
      margin: 0 auto;
      color: #fff;
      width: 140px;
      height: 30px;
      background-color: dodgerblue;
      border: 1px solid #fff;
      border-radius: 5px;
      cursor: pointer;
      transition: background-color .3s;

      &:hover {
        background-color: darken(dodgerblue, 20%);
      }
    }
  }

  & .registerButton {
    text-align: center;

    & .registerLink {
      color: dodgerblue;
      font-size: 12px;
      text-decoration: underline;
      cursor: pointer;

      &:hover {
        color: blue;
      }
    }
  }
}