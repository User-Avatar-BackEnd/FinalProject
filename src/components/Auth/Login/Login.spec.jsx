import React from 'react';
import { shallow } from 'enzyme';
import { Formik } from 'formik';

import Login from './Login';

describe('Login component', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Login />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should render Sign in form', () => {
    const formik = wrapper.find(Formik);

    expect(formik.length).toBe(1);

    expect(wrapper).toMatchSnapshot();
  });
});