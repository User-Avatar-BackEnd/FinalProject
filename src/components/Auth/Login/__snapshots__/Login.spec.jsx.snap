// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should render Sign in form 1`] = `
<div
  className="Login"
>
  <h2>
    Sign in
  </h2>
  <Formik
    initialValues={
      Object {
        "email": "",
        "password": "",
      }
    }
    onSubmit={[Function]}
    validationSchema={
      ObjectSchema {
        "_blacklist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_blacklistError": undefined,
        "_excludedEdges": Array [],
        "_nodes": Array [
          "password",
          "email",
        ],
        "_sortErrors": [Function],
        "_typeError": [Function],
        "_whitelist": ReferenceSet {
          "list": Set {},
          "refs": Map {},
        },
        "_whitelistError": undefined,
        "conditions": Array [],
        "deps": Array [],
        "exclusiveTests": Object {},
        "fields": Object {
          "email": StringSchema {
            "_blacklist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_blacklistError": undefined,
            "_mutate": undefined,
            "_typeError": [Function],
            "_whitelist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_whitelistError": undefined,
            "conditions": Array [],
            "deps": Array [],
            "exclusiveTests": Object {
              "email": false,
              "required": true,
            },
            "spec": Object {
              "abortEarly": true,
              "nullable": false,
              "presence": "required",
              "recursive": true,
              "strict": false,
              "strip": false,
            },
            "tests": Array [
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
            "type": "string",
          },
          "password": StringSchema {
            "_blacklist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_blacklistError": undefined,
            "_mutate": undefined,
            "_typeError": [Function],
            "_whitelist": ReferenceSet {
              "list": Set {},
              "refs": Map {},
            },
            "_whitelistError": undefined,
            "conditions": Array [],
            "deps": Array [],
            "exclusiveTests": Object {
              "matches": false,
              "max": true,
              "min": true,
              "required": true,
              "trim": false,
            },
            "spec": Object {
              "abortEarly": true,
              "nullable": false,
              "presence": "required",
              "recursive": true,
              "strict": false,
              "strip": false,
            },
            "tests": Array [
              [Function],
              [Function],
              [Function],
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
              [Function],
            ],
            "type": "string",
          },
        },
        "spec": Object {
          "abortEarly": true,
          "nullable": false,
          "presence": "optional",
          "recursive": true,
          "strict": false,
          "strip": false,
        },
        "tests": Array [],
        "transforms": Array [
          [Function],
        ],
        "type": "object",
      }
    }
  >
    <Component />
  </Formik>
  <div
    className="registerButton"
  >
    <Link
      className="registerLink"
      to="/registration"
    >
      Don't have an account yet? Register
    </Link>
  </div>
  <ErrorMessage
    text=""
    type="form"
  />
</div>
`;
